#2648 Spring.SendLuaUIMsg

vjikholg opened the pull request:
Implements #2648 - previous PR was on a branch that for some reason duplicated 4000+ lines of code. This is on a new branch which shouldn't have this issue. 

Spring.SendLuaUIMsg now has a single target mode for integer x 0-251, while retaining 'a', 's', '\0' compatibility. Compatible with Zero K widgets calling [SendLuaUIMsg](https://github.com/ZeroK-RTS/Zero-K/blob/b331942142d0ed9ed8e45ee9e345398aa8951f95/LuaUI/Widgets/api_selectionsend.lua#L79) as well.

vjikholg committed:
Refactor Lua UI messaging to use explicit recipient IDs instead of string modes.

The Spring.SendLuaUIMsg API now accepts either a string ("s", "a", or "") retaining compatibility or an integer (0–255) to specify recipients , supporting direct messaging to individual players as well as groups (everyone, spectators, allies).

Updated network packet handling and protocol code to match the new recipient ID scheme.

vjikholg committed:
Merge branch 'beyond-all-reason:master' into SendLuaUIMsg-work-branch

vjikholg committed:
Added enum RECIPIENT_TYPES in LuaHandle.h to reduce magic numbering (252-254). Updated LuaHandle, LuaUnsyncedCtrl to reflect these changes.

vjikholg committed:
Merge branch 'beyond-all-reason:master' into SendLuaUIMsg-work-branch

vjikholg committed:
Merge branch 'beyond-all-reason:master' into #2648-SendLuaUIMsg

vjikholg committed:
test commit transferring branches to avoid copying 4000+ lines of code during merge, unsure why code got reduplicated during merge from main -> fork

vjikholg committed:
Merge branch '#2648-SendLuaUIMsg' of https://github.com/vjikholg/RecoilEngine into #2648-SendLuaUIMsg

p2004a reviewed (commented):
Only skimming through, it seems that:
- Breaks autohost interface compatibility: can't distinguish between old and new format
- Breaks replay parsing: can't distinguish between old and new format

vjikholg commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -3690,13 +3691,34 @@ int LuaUnsyncedCtrl::SendLuaUIMsg(lua_State* L)
+
+	if (recipientID != -1) { // ensure valid recipientID
+		try {
+			clientNet->Send(CBaseNetProtocol::Get().SendLuaMsg(gu->myPlayerNum, LUA_HANDLE_ORDER_UI, recipientID, data));
```
Using this method, looking at the engineAutohostInterface.ts in this [repo](https://github.com/beyond-all-reason/recoil-autohost/blob/823a8350330b927116b75052fe78426e3c21b0f5/src/engineAutohostInterface.ts#L238), it might be necessary that autohost-side might need to push the new value into the LuaMsgScript enum, [else it will not know how to handle the packet](https://github.com/beyond-all-reason/recoil-autohost/blob/823a8350330b927116b75052fe78426e3c21b0f5/src/engineAutohostInterface.ts#L470C1-L471C1). Similarly here for the following replay parser [gex](https://github.com/Varunda/gex/blob/b492e37271d20584afd6eb7bd842f573bc3c046a/gex/Services/Parser/BarDemofileParser.cs#L554), they'd need to implement an additional value to handle the new type of LuaUIMsg. For now, I'll have to to see if there's a better solution without requiring additional work from other repositories.

sprunk replied:
I don't think it's possible to do it in a way that completely avoids any work, but a future-proof autohost or replay parser should take into account the possibility of the protocol being extended and handle anything new gracefully. In particular, [`recoil-autohost` just prints a warning when it doesn't know a packet](https://github.com/beyond-all-reason/recoil-autohost/blob/main/src/engineRunner.ts#L278-L279), and `gex` doesn't even look at the script or mode parameters.

We should probably be nice and use the existing format where applicable so as not to introduce a regression (i.e. specific playerID would be the new script arg but allies/spectators/everyone would be kept as-is for now) and give people some time to adjust, but other than that it's fine to just add new things as long as existing values don't change their meaning.

p2004a replied:
As long as update won't be butchered https://discord.com/channels/549281623154229250/549282166543089674/1430649298445406350 changes by charging e.g. message id could be fine in such cases (would also cover autohost because protocol message id is uniquely copied also to autohost for this message) but what is proposed so extension on size also makes it possible to detect.

We had idea last time how to make changes to network protocol by incrementing a new field in header: https://discord.com/channels/223585969956323328/955007493510160384/1401303838019027126 and I asked to do the same by adding new int to first packet `SERVER_STARTED` to autohost and incrementing on changes like this.

lhog replied:
I was going to explore https://flatbuffers.dev/ , but so far never had enough drive to push it through.

vjikholg committed:
added LUA_HANDLE_ORDER_UI_SINGLE = 2100 (LUA_HANDLE_ORDER_UI = 2000). Allows replay parsers/autohost to differentiate between packets. Updated HandleLuaMsg to reflect these changes. Swapped SendLuaUIMsg to use old logic upon receiving string, new logic upon receiving int.

vjikholg committed:
added LUA_HANDLE_ORDER_UI_SINGLE = 2100 (LUA_HANDLE_ORDER_UI = 2000). Allows replay parsers/autohost to differentiate between packets. Updated HandleLuaMsg to reflect these changes.

vjikholg committed:
Merge branch 'beyond-all-reason:master' into #2648-SendLuaUIMsg

vjikholg committed:
Main -> Fork merge

p2004a commented:
Stepping back a bit, what is even the use case of #2648 @sprunk? What game wants to do what?

If you squint, you could consider #2040 to be related and in https://github.com/beyond-all-reason/RecoilEngine/pull/2355#issuecomment-2955679139 we considered pushing that through `LuaUIMsg`, but with preference to use dedicated network message in the end.

One note from that investigation maybe is also relevant here:

> The Callins:RecvLuaMsg also just takes player and message, the mode is not forwarded

so, upon receiving the lua msg callin, game has no way to know whatever that luamsg was send to everobydy, allies, only that player. Is that a problem for whatever use case this functionality is supposed to be used for?

p2004a reviewed (changes requested):
Thanks Sprung, good points and I agree, use case seems reasonable.

To the PR itself, the constructions optimizes for backwards compatibility, not clarity, maybe that's still good call at this point, and let's see how things evolve.

vjikholg committed:
Merge branch 'beyond-all-reason:master' into #2648-SendLuaUIMsg

vjikholg committed:
Updated + Hardened LuaunsyncedCtrl::SendLuaUIMsg to tolerate invalid types w/ better error handling, explicitly added case where 2nd argument not received. Also updated descriptor wording + formatting.

Readded LuaHandle::LUAMSG_TYPES enum to solve magic numbering for HandleLuaMsg

Renamed HandleLuaMsg variable to better reflect usage, uses LUAMSG_TYPES enum to avoid magic numbering within the LUA_HANDLE_ORDER_UI case.

Renamed CBaseNetProtocol::SendLuaMsg mode argument to recipientID to better reflect usage.

vjikholg committed:
Merge branch '#2648-SendLuaUIMsg' of https://github.com/vjikholg/RecoilEngine into #2648-SendLuaUIMsg

vjikholg committed:
Fixed variable naming typo inside CBaseNetProtocol::SendLuaMsg

p2004a commented on rts/Lua/LuaHandle.cpp:
```diff
@@ -2419,35 +2420,44 @@ void CLuaHandle::SwapEnableModule(lua_State* L, bool enabled, const char* module
 							const int msgAllyTeam = teamHandler.AllyTeam(player->team);
 							sendMsg = teamHandler.Ally(msgAllyTeam, gu->myAllyTeam);
-						}
+						} 
```
```suggestion
						}
```

vjikholg committed:
Fixes descriptor formatting/whitespace in `CBaseNetProtocol::SendLuaMsg`

Fixes function descriptor formatting + error typo for LuaUnsyncedCtrl::SendLuaUIMsg

Fixes logic formatting and whitespace in LuaHandle::HandleLuaMsg.

Fixes enum LUAMSG_TYPES implicitly casting `uint8 -> int` by explicitly setting enum-value type in declaration.

Co-authored-by: Marek Rusinowski <marekrusinowski@gmail.com>

vjikholg committed:
Update rts/Net/Protocol/BaseNetProtocol.cpp

Co-authored-by: Marek Rusinowski <marekrusinowski@gmail.com>

vjikholg committed:
Merge branch '#2648-SendLuaUIMsg' of https://github.com/vjikholg/RecoilEngine into #2648-SendLuaUIMsg

sprunk commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -3679,23 +3679,41 @@ int LuaUnsyncedCtrl::SetBuildFacing(lua_State* L)
+		scriptID = LUA_HANDLE_ORDER_UI;
+
+	} else if (lua_israwnumber(L, 2)) {
+		recipientID = luaL_optnumber(L, 2, -1); // if no arg, outputs -1
```
Can't be no arg since it was just checked in the line above.
```suggestion
	} else if (lua_israwnumber(L, 2)) {
		recipientID = lua_tonumber(L, 2);
```

sprunk commented on rts/Lua/LuaUnsyncedCtrl.cpp:
```diff
@@ -3679,23 +3679,41 @@ int LuaUnsyncedCtrl::SetBuildFacing(lua_State* L)
+	} else if (lua_israwnumber(L, 2)) {
+		recipientID = luaL_optnumber(L, 2, -1); // if no arg, outputs -1
+		if (recipientID < 0 || recipientID > MAX_PLAYERS) {
+			luaL_error(L, "Invalid player ID sent (expecting between 0-251): ", recipientID);
```
Needs to use printf formatting, and also use the magic constant rather than hardcode 251: 
```suggestion
			luaL_error(L, "Invalid player ID sent (%d, expecting between 0-%d): ", recipientID, (int) MAX_PLAYERS);
```

Ditto for the other error calls.

vjikholg committed:
Merge branch 'beyond-all-reason:master' into #2648-SendLuaUIMsg

vjikholg committed:
Fixes LuaUnsyncedCtrl::SendLuaUIMsg error messages, now properly uses printf formatting when yielding verbose errors.

vjikholg committed:
Moves mode pointer within string case for SendLuaUiMsg

vjikholg committed:
Merge branch 'beyond-all-reason:master' into #2648-SendLuaUIMsg
